File: app/src/androidTest/java/co/edu/uniandes/miswmobile/vinilosapp/AlbumsListTest.java
package co.edu.uniandes.miswmobile.vinilosapp;
import androidx.test.espresso.ViewInteraction;
import androidx.test.filters.LargeTest;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.example.vinyls_jetpack_application.R;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Timeout;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.replaceText;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import co.edu.uniandes.miswmobile.vinilosapp.ui.AccessActivity;
import kotlin.jvm.Throws;

@RunWith(AndroidJUnit4.class)
public class AlbumsListTest {

    @Rule
    public ActivityScenarioRule<AccessActivity> mActivityTestRule = new ActivityScenarioRule<>(AccessActivity.class);
    

    @Test
    public void mainActivityAlbumsTest() {
        ViewInteraction buttonVisitor = onView(allOf(withId(R.id.buttonVisitor), withText("Visitante"), isDisplayed()));
        buttonVisitor.perform(click());

        ViewInteraction buttonAlbums = onView(allOf(withId(R.id.buttonAlbums), withText("Álbumes"), isDisplayed()));
        buttonAlbums.perform(click());

        onView(allOf(withId(R.id.action_bar), withText("Álbumes"), isDisplayed()));

        onView(allOf(withId(R.id.albumsRv), isDisplayed()));


    }
}

File: app/src/androidTest/java/co/edu/uniandes/miswmobile/vinilosapp/AlbumsListTest.kt
package co.edu.uniandes.miswmobile.vinilosapp

import android.widget.Spinner
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.IdlingRegistry
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.assertThat
import androidx.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import co.edu.uniandes.miswmobile.vinilosapp.ui.AccessActivity
import org.hamcrest.Matchers.greaterThan
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class AlbumsListTest {

    @Rule
    @JvmField
    var mActivityTestRule = ActivityScenarioRule(
        AccessActivity::class.java
    )

    private val idlingResource = SimpleIdlingResource("albumsRv")

    @Before
    fun setUp() {
        IdlingRegistry.getInstance().register(idlingResource)
    }

    @After
    fun tearDow() {
        IdlingRegistry.getInstance().unregister(idlingResource)
    }

    @Test
    fun listAlbumsAsVisitorTest() {

        var itemCount: Int = 0
        var randomItemAlbum: Int = -1

        val toolbar2 = onView(withId(R.id.toolbar2))
        toolbar2.check(matches(isDisplayed()))

        val buttonVisitor = onView(withId(R.id.buttonVisitor))
        buttonVisitor.check(matches(isDisplayed()))
        buttonVisitor.perform(click())

        var actionBar = onView(withId(R.id.my_toolbar))
        actionBar.check(matches(isDisplayed()))

        val buttonAlbums = onView(withId(R.id.buttonAlbums))
        buttonAlbums.check(matches(isDisplayed()))
        buttonAlbums.perform(click())

        actionBar = onView(withId(R.id.my_toolbar))
        actionBar.check(matches(isDisplayed()))
        val rvAlbums = onView(withId(R.id.albumsRv))
        rvAlbums.check(matches(isCompletelyDisplayed()))

        // obtener numero de items mostrados, debe ser mayor a cero
        rvAlbums.check { view, _ ->
            val recyclerView = view as RecyclerView
            itemCount = recyclerView.adapter!!.itemCount
        }
        assertThat(itemCount, greaterThan(0))

    }

    @Test
    fun listAlbumsAsCollectorTest() {
        var itemCount: Int = 0
        var randomItemAlbum: Int = -1

        val toolbar2 = onView(withId(R.id.toolbar2))
        toolbar2.check(matches(isDisplayed()))

        val buttonCollectors = onView(withId(R.id.buttonCollector))
        buttonCollectors.check(matches(isDisplayed()))

        var randomItem = -1
        var itemSelected: Any = ""
        onView(withId(R.id.spinner)).check { view, _ ->
            val appCompatSpinner = view as Spinner
            assertThat(appCompatSpinner.count, greaterThan(0))
            randomItem = (0 until appCompatSpinner.count).random()
            itemSelected = appCompatSpinner.getItemAtPosition(randomItem)
        }
        onView(withId(R.id.spinner)).perform(click())
        onView(withText(itemSelected.toString())).perform(click())

        var actionBar = onView(withId(R.id.my_toolbar))
        actionBar.check(matches(isDisplayed()))

        val buttonAlbums = onView(withId(R.id.buttonAlbums))
        buttonAlbums.check(matches(isDisplayed()))
        buttonAlbums.perform(click())

        actionBar = onView(withId(R.id.my_toolbar))
        actionBar.check(matches(isDisplayed()))
        val rvAlbums = onView(withId(R.id.albumsRv))
        rvAlbums.check(matches(isCompletelyDisplayed()))

        // verificar boton de accion agregar album este presente
        val buttonAddAlbum = onView(withId(R.id.add_album))
        buttonAddAlbum.check(matches(isDisplayed()))

        // obtener numero de items mostrados, debe ser mayor a cero
        rvAlbums.check { view, _ ->
            val recyclerView = view as RecyclerView
            itemCount = recyclerView.adapter!!.itemCount
        }
        assertThat(itemCount, greaterThan(0))
    }

}

